### PR TITLE
fixing pytest: File with syntax error prevents discovery of tests  #5471

### LINKED ISSUE #5471: pytest: File with syntax error prevents discovery of tests
**Steps to Reproduce**
1. Create a project with 2 files
2. In one file, have a valid pytest
3. In the other file, have a syntax error
4. Close the solution and reopen
5. Wait for tests discovery to complete

Because of the file with syntax error, none of the tests in the project are discovered.

**Expected behavior**
Test is detected an appears in solution explorer
When a file has syntax error, it doesn't break the rest of discovery

See `DiscoverPytestSyntaxError` test for a repro. That test should pass once bug is fixed (the test is currently ignored)

### LINKED ISSUE #5529: Python warnings breaks test discovery using unittest framework
Note: this only applies to unittest, pytest works fine in this situation

Create new project
Delete the main python application file
Add new unit test
Add this code at the top of the test file:
```
import warnings
warnings.warn('you have been warned!')
```

Enable test using unittest framework

Result: test discovery failure

```
[8/2/2019 8:44:02.887 AM Informational] ---------- Discovery started ----------
[8/2/2019 8:44:03.921 AM Informational] Starting unittest discovery for the PythonApplication6 project. This can take up to 60 seconds.
[8/2/2019 8:44:04.418 AM Error] Failed to parse: Unexpected character encountered while parsing value: C. Path '', line 0, position 0.
[8/2/2019 8:44:04.419 AM Error] C:\Users\huvalo\source\repos\PythonApplication6\test1.py:4: UserWarning: you have been warned!
  warnings.warn('you have been warned!')
[{"tests": [{"id": "test1.Test_test1.test_A", "name": "test_A", "lineno": 7, "source": "C:\\Users\\huvalo\\source\\repos\\PythonApplication6\\test1.py"}]}]

[8/2/2019 8:44:04.420 AM Error] Object reference not set to an instance of an object.
[8/2/2019 8:44:04.423 AM Warning] No test is available in c:\users\huvalo\source\repos\pythonapplication6\test1.py. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
[8/2/2019 8:44:04.505 AM Informational] ========== Discovery finished: 0 tests found (0:00:01.579745) ==========
```

Some info on warnings here and how to disable them
https://www.idiotinside.com/2016/12/17/python-warnings-framework/

For example, adding this might work:
```
warnings.simplefilter('ignore')
```

### LINKED ISSUE #5552: Unittest discovery errors are not being shown in test output window
**Describe the bug**

**Steps to Reproduce**
1.  Add a syntax error to a unittest file ie. "]"
2.
3.

**Expected behavior**
Traceback (most recent call last):
  File "C:\Users\bschnurr\AppData\Local\Programs\Python\Python37-32\lib\unittest\loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "C:\Users\bschnurr\AppData\Local\Programs\Python\Python37-32\lib\unittest\loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "C:\Users\bschnurr\source\repos\PythonApplication2\PythonApplication2\test2.py", line 9
    ]
    ^
SyntaxError: invalid syntax

### PR DESCRIPTION
**Pull request by bschnurr**
pytest: File with syntax error prevents discovery of tests Fix #5471
Python warnings breaks test discovery using unittest framework Fix #5529
Unittest discovery errors are not being shown in test output window Fix #5552

- Previously both errors and results were being written to stdout in the python discovery adapter.
- Now adapter script takes an optional outputfile to write discovery results to.
- Fixed the script so that it still reports items found even if there is an error in pytest
- Added unit tests around syntax errors still reporting partial results and error logging is being reported now by unittest discovery

Note: Issue still remaining with unittest discovery in python 2.7 not reporting errors to stdout.

### CODE REVIEW

File: Python/Product/Common/Strings.resx [hunk 1]
```diff
@@ -2935,4 +2935,7 @@ Please specify at least one package.</value>
     <value>A file containing Python tests was detected in the {0} {1}, but the test framework is disabled.</value>
     <comment>{0} is project or workspace name. {1} is "project" or "workspace"</comment>
   </data>
+  <data name="PythonDiscoveryResultsNotFound" xml:space="preserve">
+    <value>Discovery results file '{0}' was not found {</value>
```
Comment (huguesv) [inline Python/Product/Common/Strings.resx]: there's a extra curly brace at the end (and missing period)

File: Python/Product/TestAdapter.Executor/Pytest/TestDiscovererPytest.cs [hunk 2]
```diff
@@ -40,53 +40,63 @@ internal class TestDiscovererPytest : IPythonTestDiscoverer {
...
+            }
+
             List<PytestDiscoveryResults> results = null;
             try {
                 results = JsonConvert.DeserializeObject<List<PytestDiscoveryResults>>(json);
+                CreateVsTests(results, discoverySink);
             } catch (InvalidOperationException ex) {
                 Error("Failed to parse: {0}".FormatInvariant(ex.Message));
-                Error(json);
+                Error(json ?? string.Empty);
```
Comment (huguesv) [inline Python/Product/TestAdapter.Executor/Pytest/TestDiscovererPytest.cs]: impossible for it to be null, same below

File: Python/Product/TestAdapter.Executor/Pytest/TestDiscovererPytest.cs [hunk 3]
```diff
@@ -40,53 +40,63 @@ internal class TestDiscovererPytest : IPythonTestDiscoverer {
...
             LogInfo(Strings.PythonTestDiscovererStartedMessage.FormatUI(PythonConstants.PytestText, _settings.ProjectName, workspaceText, _settings.DiscoveryWaitTimeInSeconds));
 
-            try {
-                var env = InitializeEnvironment(sources, _settings);
-                var arguments = GetArguments(sources, _settings);
-                DebugInfo("cd " + _settings.WorkingDirectory);
-                DebugInfo("set " + _settings.PathEnv + "=" + env[_settings.PathEnv]);
-                DebugInfo($"{_settings.InterpreterPath} {string.Join(" ", arguments)}");
+            var env = InitializeEnvironment(sources, _settings);
+            var outputfilename = Path.GetTempFileName();
```
Comment (huguesv) [inline Python/Product/TestAdapter.Executor/Pytest/TestDiscovererPytest.cs]: prefer camel case: `outputFilePath` or `outputFileName`

File: Python/Product/TestAdapter.Executor/UnitTest/TestDiscovererUnitTest.cs [hunk 4]
```diff
@@ -54,63 +54,75 @@ internal class TestDiscovererUnitTest : IPythonTestDiscoverer {
...
             var workspaceText = _settings.IsWorkspace ? Strings.WorkspaceText : Strings.ProjectText;
             LogInfo(Strings.PythonTestDiscovererStartedMessage.FormatUI(PythonConstants.UnitTestText, _settings.ProjectName, workspaceText, _settings.DiscoveryWaitTimeInSeconds));
 
-            try {
-                var env = InitializeEnvironment(sources, _settings);
-                var arguments = GetArguments(sources);
-                DebugInfo("cd " + _settings.WorkingDirectory);
-                DebugInfo("set " + _settings.PathEnv + "=" + env[_settings.PathEnv]);
-                DebugInfo($"{_settings.InterpreterPath} {string.Join(" ", arguments)}");
+            var env = InitializeEnvironment(sources, _settings);
```
Comment (huguesv) [inline Python/Product/TestAdapter.Executor/UnitTest/TestDiscovererUnitTest.cs]: same comments as in the file above

File: Python/Product/Common/Strings.resx (hunk 1 above)
Comment (bschnurr) [reply Python/Product/Common/Strings.resx]: should i add a period too?

Comment (huguesv) [reply Python/Product/Common/Strings.resx]: yeah

**Pull request merged**